tests/test_sample_rayservice_yamls.py

```diff
@@ -1,55 +1,338 @@
+                    rulesets=[RuleSet([EasyJobRule(), CurlServiceRule(queries=updated_queries)])],
+                    timeout=90,
+                    namespace=NAMESPACE,
+                    filepath=yaml_copy.name
+                ),
+                RayServiceDeleteCREvent(self.cr, [], 90, NAMESPACE, self.sample_path),
+            ]
+
+            for cr_event in cr_events:
+                cr_event.trigger()
+
+    def test_zero_downtime_rollout(self, set_up_cluster):
+        # Modify the cluster spec to trigger a rollout
+        updated_cr = deepcopy(self.cr)
+
+        config = yaml.safe_load(self.cr["spec"]["serveConfigV2"])
+        config["applications"][0]["deployments"][0]["user_config"]["price"] = 4
+        config["applications"][1]["deployments"][1]["user_config"]["factor"] = 3
+        updated_cr["spec"]["serveConfigV2"] = yaml.safe_dump(config)
+
+        env = [{"name": "SAMPLE_ENV_VAR", "value": "SAMPLE_VALUE"}]
+        updated_cr["spec"]["rayClusterConfig"]["headGroupSpec"]["template"]["spec"]["containers"][0]["env"] = env
+
+        updated_queries = [
+            {"path": "/fruit", "json_args": ["MANGO", 2], "expected_output": "8"},
+            {"path": "/calc", "json_args": ["MUL", 3], "expected_output": '"9 pizzas please!"'},
+        ]
+        allowed_queries_during_update = deepcopy(self.default_queries)
+        allowed_queries_during_update[0]["expected_output"] = {"6", "8"}
+        allowed_queries_during_update[1]["expected_output"] = {'"15 pizzas please!"', '"9 pizzas please!"'}
+
+        with NamedTemporaryFile(mode="w+", suffix=".yaml") as yaml_copy:
+            logger.info(f"Writing modified RayService yaml to {yaml_copy.name}.")
+            yaml_copy.writelines(yaml.safe_dump(updated_cr))
+            yaml_copy.flush()
+
+            cr_events: List[CREvent] = [
+                RayServiceAddCREvent(
+                    custom_resource_object=self.cr,
+                    rulesets=[RuleSet([EasyJobRule(), CurlServiceRule(queries=self.default_queries)])],
+                    filepath=self.sample_path
+                ), 
+                RayServiceUpdateCREvent(
+                    custom_resource_object=self.cr,
+                    rulesets=[RuleSet([CurlServiceRule(queries=updated_queries)])],
+                    filepath=yaml_copy.name,
+                    switch_cluster=True,
+                    query_while_updating=allowed_queries_during_update,
+                ), 
+                RayServiceDeleteCREvent(custom_resource_object=self.cr, filepath=self.sample_path),
+            ]
+
+            for cr_event in cr_events:
+                cr_event.trigger()
+
+    def test_service_autoscaling(self, set_up_cluster):
+        filename = "ray-service.autoscaler.yaml"
+        path = CONST.REPO_ROOT.joinpath("ray-operator/config/samples/").joinpath(filename)
+
+        scale_up_rule = AutoscaleRule(
+            query={"path": "/", "json_args": {}},
+            num_repeat=20,
```

Comment:
Does it mean sending 20 requests? How do we know it will scale up to 14 replicas (7 CPUs) instead of another number, such as 5 replicas?

Reply:
I do not understand when will the Serving autoscaling be triggered.

Reply:
I use the deployment from https://github.com/ray-project/serve_workloads/blob/main/autoscaling_test/blocked.py. If you send requests to this deployment, the request will block indefinitely until https://github.com/ray-project/serve_workloads/blob/main/autoscaling_test/signaling.py releases the "lock". The target number of requests ongoing requests per replica is 1, so serve will try to add enough replicas to serve all 20 requests (since all requests are blocked).

Reply:
Thank you for the explanation! We may need to add comments to both this test and the YAML file.

Reply:
Good point. I've added comments to both the test and yaml.